### PR TITLE
feat(goctl): supports api  multi-level importing

### PR DESCRIPTION
**Pull request by fynxiu**
Resolves: #1744

### CODE REVIEW

File: tools/goctl/api/parser/g4/ast/apiparser.go [hunk 1]
```diff
@@ -184,58 +207,40 @@ func (p *Parser) invoke(linePrefix, content string) (v *Api, err error) {
...
-	mainHandlerMap := make(map[string]PlaceHolder)
-	mainRouteMap := make(map[string]PlaceHolder)
-	mainTypeMap := make(map[string]PlaceHolder)
-
-	routeMap := func(list []*ServiceRoute) (map[string]PlaceHolder, map[string]PlaceHolder) {
-		handlerMap := make(map[string]PlaceHolder)
-		routeMap := make(map[string]PlaceHolder)
-
+// storeVerificationInfo stores information for verification
+func (p *Parser) storeVerificationInfo(api *Api) {
```
Comment (anqiansong) [inline tools/goctl/api/parser/g4/ast/apiparser.go:211]: In this code block, the method `p.checkTypeDeclaration` should be called before put a `handler` & `route` into a map, by this way, the duplicate declarations error should be appeared and give a feedback to user.

**Review by anqiansong — COMMENTED** [review]
The method `p.nestedApiCheck` should be removed because it's useless.

File: tools/goctl/api/parser/g4/ast/apiparser.go [hunk 2]
```diff
@@ -113,7 +119,27 @@ func (p *Parser) parse(filename, content string) (*Api, error) {
 
 	var apiAstList []*Api
 	apiAstList = append(apiAstList, root)
-	for _, imp := range root.Import {
+	p.storeVerificationInfo(root)
+	impApiAstList, err := p.invokeImportedApi(root.Import)
```
Comment (anqiansong) [inline tools/goctl/api/parser/g4/ast/apiparser.go:123]: It is possible that a circular import error will occur, maybe we should fix it in this pr or next pr.

Comment (fynxiu) [reply tools/goctl/api/parser/g4/ast/apiparser.go:123]: I'll fix it.

File: tools/goctl/api/parser/g4/ast/importstack.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package ast
+
+import "errors"
+
+var ErrImportCycleNotAllowed = errors.New("import cycle not allowed")
```
Comment (kevwan) [inline tools/goctl/api/parser/g4/ast/importstack.go:5]: Exported vars and methods need to be commented to pass linter.

**Review by kevwan — APPROVED** [review]
lgtm.

File: tools/goctl/api/parser/g4/ast/importstack.go (hunk 3 above)
Comment (fynxiu) [reply tools/goctl/api/parser/g4/ast/importstack.go:6]: Thanks, fixed it.

File: tools/goctl/api/parser/g4/ast/apiparser.go [hunk 4]
```diff
@@ -113,12 +125,43 @@ func (p *Parser) parse(filename, content string) (*Api, error) {
...
 		dir := filepath.Dir(p.src)
 		impPath := strings.ReplaceAll(imp.Value.Text(), "\"", "")
 		if !filepath.IsAbs(impPath) {
 			impPath = filepath.Join(dir, impPath)
 		}
+		// import cycle check
+		if err := p.importStatck.push(impPath); err != nil {
+			return nil, err
+		}
+		// ignore alread imported file
```
Comment (anqiansong) [inline tools/goctl/api/parser/g4/ast/apiparser.go:159]: alread -> already

**Review by marlongerson — APPROVED** [review]

File: tools/goctl/api/parser/g4/ast/apiparser.go (hunk 1 above)
Comment (fynxiu) [reply tools/goctl/api/parser/g4/ast/apiparser.go:234]: Hi, @anqiansong, p.valid has indeed been invoked in `apiparser.go` at line 175.

**Review by anqiansong — APPROVED** [review]

**Comment by waltcow** [comment]
Issue still not resoved in latest goctl version
 the nested api does not support import

```go
//a.api
import "./b.api"

type A {
    b B `json:"b"`
}

```

```go
// b.api
type B {

}
```


 goctl env                      
```

GOCTL_OS=darwin
GOCTL_ARCH=amd64
GOCTL_HOME=/Users/xxx/.goctl
GOCTL_DEBUG=False
GOCTL_CACHE=/Users/v/.goctl/cache
GOCTL_VERSION=1.3.4
PROTOC_VERSION=3.19.1
PROTOC_GEN_GO_VERSION=v1.27.1
PROTO_GEN_GO_GRPC_VERSION=1.2.0


```

**Pull request merged**